### DISCUSSION TITLE
Restrict function for specific HTTP methods

### DISCUSSION THREAD

**Discussion by StefMa** [vercel/vercel#6438] [Help]
Hi folks 👋 

I'm kind of new in the backend-scene and all the "how rest works".
But I have the following use-case:
My API should have two "one route" but I want to have invoke different functions, depending on the HTTP method.

For example:
`/api/game` POST -> `createNewGame.go`
`/api/game` GET -> `getGame.go`

How can I do this?
As far as I see it, it could work with the *legacy* [`routes` configuration](https://vercel.com/docs/configuration#project/routes).
But, well, I don't want to start with something which is marked as *legacy* 😁 .

The only possible solution would be, that I create just "one handler" for `/api/game`, check the HTTP method and, depending on it, call another file. But this sounds for some reasons strange to some...

Any other ideas?
Did I miss something?

Thanks in advance!

**Comment by mcsdevv** [comment]
Hi @StefMa, I'd recommend using the built-in helps with `req.method` inside an `if` statement. You can then respond to different HTTP methods conditionally. Further information is available in the [documentation](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-request-and-response-objects/node-js-helpers).

**Reply by StefMa** [reply]
Thanks!
I use Go, but I guess this doesn't matter 😁 .
Thought there is something build it by `vercel`.
But I will go this way.

**Comment by barelyhuman** [comment]
The overall go code will get bundled into one single binary called the entrypoint.go which is what the function will execute. 

If you wish to have 2 files specifically for the reason to maintain separation in the codebase, you can use one entry file with 2 separate function just like @mcsdevv  suggested.

Which goes like 
```go

func handler(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodGet {
		io.WriteString(w, "This is a get request")
	} else if r.Method == http.MethodPost {
		io.WriteString(w, "This is a post request")
	}
}
```

but since you want it to be in different files, what you can instead do is use go modules , you'll have to be on version 1.11 or greater for these to work. 

You could follow along this answer to handle the helpers in other files and then use the defined functions here in the if/else above 
https://github.com/vercel/vercel/discussions/6316

Hope that helps

**Reply by StefMa** [reply]
Hi,

thanks for the suggestions.
I already thought about a similar architecture while reading @mcsdevv answer a few days ago.

I will probably end up with one handler for each route:
```
api/game.go
api/somethingElse.go
```
But each of them will nothing else than:
* checking the HTTP method
* calling my "controller" code (depending on the HTTP method)
```
// example api/game.go

import (
   ...
  my.go.module/controller
)

func handler(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodGet {
		controller.getGame(w, r)
	} else if r.Method == http.MethodPost {
		controller.createGame(w, r)
	}
}
```

My controller package could then have different files:
```
// controller/getGame.go
func getGame(w http.ResponseWriter, r *http.Request) { }

// controller/createGame.go
func createGame(w http.ResponseWriter, r *http.Request) { }
```

With that solution I woul have even better separation of concers I guess...

**Reply by barelyhuman** [reply]
Basically what I was trying to suggest  😂 

Yeah that should be fine, unless you are building a super complex api, this approach should be fine. 

If you start involving Databases and remote connections that could cross the 10 second limit on Vercel , in those cases you are better of handling the connections and all in the actual api endpoint instead of individual handler and pass them to the handler instead.
